src/dfx/src/commands/extension/list.rs

```diff
@@ -1,22 +1,67 @@
 use crate::lib::environment::Environment;
 use crate::lib::error::DfxResult;
+use clap::Parser;
 use std::io::Write;
+use tokio::runtime::Runtime;
+use url::Url;
 
-pub fn exec(env: &dyn Environment) -> DfxResult<()> {
+#[derive(Parser)]
+pub struct ListOpts {
+    /// Specifies to list the available remote extensions.
+    #[arg(long)]
+    available: bool,
+    /// Specifies the URL of the catalog to use to find the extension.
+    #[clap(long)]
+    catalog_url: Option<Url>,
+}
+
+pub fn exec(env: &dyn Environment, opts: ListOpts) -> DfxResult<()> {
     let mgr = env.get_extension_manager();
-    let extensions = mgr.list_installed_extensions()?;
+    let extensions;
+
+    let result;
+    if opts.available || opts.catalog_url.is_some() {
+        let runtime = Runtime::new().expect("Unable to create a runtime");
+        extensions = runtime.block_on(async {
+            mgr.list_available_extensions(opts.catalog_url.as_ref())
+                .await
+        })?;
+
+        result = display_extension_list(
+            &extensions,
+            "No extensions available.",
+            "Available extensions:",
+        );
+    } else {
+        extensions = mgr.list_installed_extensions()?;
 
+        result = display_extension_list(
+            &extensions,
+            "No extensions installed.",
+            "Installed extensions:",
+        );
+    };
+
+    result
```

Comment:
There's a little streamlining possible here.  `let <variable>;` is useful sometimes, but isn't needed here.

```suggestion
    if opts.available || opts.catalog_url.is_some() {
        let runtime = Runtime::new().expect("Unable to create a runtime");
        let extensions = runtime.block_on(async {
            mgr.list_available_extensions(opts.catalog_url.as_ref())
                .await
        })?;

        display_extension_list(
            &extensions,
            "No extensions available.",
            "Available extensions:",
        )
    } else {
        let extensions = mgr.list_installed_extensions()?;

        display_extension_list(
            &extensions,
            "No extensions installed.",
            "Installed extensions:",
        )
    }
```

Reply:
Fxied at https://github.com/dfinity/sdk/pull/3896/commits/7dc4dfbc673101bb2ff596e56d4a1bc9f3956763.